src/hotspot/share/code/nmethod.cpp

```diff
@@ -2596,7 +2596,7 @@ void nmethod::metadata_do(MetadataClosure* f) {
 // Main purpose is to reduce code cache pressure and get rid of
 // nmethods that don't seem to be all that relevant any longer.
 bool nmethod::is_cold() {
-  if (!MethodFlushing || is_native_method() || is_not_installed()) {
+  if (!MethodFlushing || (is_native_method() && is_in_use()) || is_not_installed()) {
```

Comment:
So I guess we need to decide what to do about native wrappers that are still "in use", but are "cold" because they haven't been called in a while.  The above change would keep them around forever.  We could instead allow them to be cleaned up like regular nmethods.

Reply:
I am tempted to say yes, for consistency, but it probably won't make much of a difference either way.  But now I am wondering, if these cold native wrappers continue to be immortal, then do they really need to give them nmethod entry barriers?  Removing the barrier could remove some overhead.  Whatever direction we decide to go, it would be good to add a comment here explaining the decision and/or trade-offs.

Reply:
https://bugs.openjdk.org/browse/JDK-8370472